File: packages/shared/test/evaluator.test.js
import assert from "node:assert/strict";
import test from "node:test";

import {
  EvaluationReason,
  bucketUser,
  evaluateFlag,
  matchesRule
} from "../src/evaluator.js";

test("missing flag returns caller default value", () => {
  const result = evaluateFlag(null, { id: "user_123" }, {
    flagKey: "new-checkout",
    defaultValue: true
  });

  assert.equal(result.enabled, true);
  assert.equal(result.reason, EvaluationReason.FLAG_NOT_FOUND);
});

test("missing flag defaults to false when caller default is not provided", () => {
  const result = evaluateFlag(null, { id: "user_123" }, {
    flagKey: "new-checkout"
  });

  assert.equal(result.enabled, false);
  assert.equal(result.reason, EvaluationReason.FLAG_NOT_FOUND);
});

test("disabled flag returns false even when targeting rule would match", () => {
  const result = evaluateFlag({
    key: "new-checkout",
    enabled: false,
    targetingRules: [
      { attribute: "country", operator: "equals", value: "ID" }
    ]
  }, {
    id: "user_123",
    country: "ID"
  });

  assert.equal(result.enabled, false);
  assert.equal(result.reason, EvaluationReason.FLAG_DISABLED);
});

test("targeting rules use OR semantics", () => {
  const result = evaluateFlag({
    key: "new-checkout",
    enabled: true,
    targetingRules: [
      { attribute: "country", operator: "equals", value: "SG" },
      { attribute: "plan", operator: "equals", value: "premium" }
    ]
  }, {
    id: "user_123",
    country: "ID",
    plan: "premium"
  });

  assert.equal(result.enabled, true);
  assert.equal(result.reason, EvaluationReason.MATCHED_RULE);
  assert.deepEqual(result.matchedRule, {
    attribute: "plan",
    operator: "equals",
    value: "premium"
  });
});

test("missing user attribute does not match a rule", () => {
  assert.equal(
    matchesRule({ attribute: "country", operator: "equals", value: "ID" }, {}),
    false
  );
});

test("supports first-version targeting operators", () => {
  const user = {
    email: "alice@company.com",
    country: "ID",
    plan: "premium"
  };

  assert.equal(matchesRule({ attribute: "country", operator: "equals", value: "ID" }, user), true);
  assert.equal(matchesRule({ attribute: "country", operator: "not_equals", value: "SG" }, user), true);
  assert.equal(matchesRule({ attribute: "email", operator: "contains", value: "alice" }, user), true);
  assert.equal(matchesRule({ attribute: "email", operator: "ends_with", value: "@company.com" }, user), true);
  assert.equal(matchesRule({ attribute: "plan", operator: "in", value: ["premium", "enterprise"] }, user), true);
});

test("percentage rollout is deterministic for the same flag and user", () => {
  const first = bucketUser("new-checkout", "user_123");
  const second = bucketUser("new-checkout", "user_123");

  assert.equal(first, second);
});

test("percentage rollout keeps enabled users when percentage increases", () => {
  const flag = {
    key: "new-checkout",
    enabled: true,
    rolloutPercentage: 10
  };
  const user = { id: "user_123" };
  const resultAt10 = evaluateFlag(flag, user);
  const resultAt25 = evaluateFlag({ ...flag, rolloutPercentage: 25 }, user);

  if (resultAt10.enabled) {
    assert.equal(resultAt25.enabled, true);
  }
});

test("zero rollout falls back to caller default", () => {
  const result = evaluateFlag({
    key: "new-checkout",
    enabled: true,
    rolloutPercentage: 0
  }, {
    id: "user_123"
  }, {
    defaultValue: false
  });

  assert.equal(result.enabled, false);
  assert.equal(result.reason, EvaluationReason.DEFAULT_RULE);
});


test("100 percent rollout enables any user with an id", () => {
  const result = evaluateFlag({
    key: "new-checkout",
    enabled: true,
    rolloutPercentage: 100
  }, {
    id: "user_123"
  });

  assert.equal(result.enabled, true)
  assert.equal(result.reason, EvaluationReason.PERCENTAGE_ROLLOUT)
} )

File: services/config-service/internal/flags/evaluator_test.go
package flags

import "testing"

func TestEvaluateMissingFlagReturnsDefaultValue(t *testing.T) {
	result := Evaluate(nil, "new-checkout", map[string]any{}, true)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonFlagNotFound {
		t.Fatalf("expected reason %s, got %s", ReasonFlagNotFound, result.Reason)
	}
}

func TestEvaluateDisabledFlagReturnsFalse(t *testing.T) {
	flag := Flag{
		Key:     "new-checkout",
		Enabled: false,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{}, true)

	if result.Enabled {
		t.Fatalf("expected enabled false")
	}

	if result.Reason != ReasonFlagDisabled {
		t.Fatalf("expected reason %s, got %s", ReasonFlagDisabled, result.Reason)
	}
}

func TestEvaluateEnabledFlagReturnsDefaultRule(t *testing.T) {
	flag := Flag{
		Key:     "new-checkout",
		Enabled: true,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonDefaultRule {
		t.Fatalf("expected reason %s, got %s", ReasonDefaultRule, result.Reason)
	}
}

func TestEvaluateMatchedCountryRuleReturnsTrue(t *testing.T) {
	flag := Flag{
		Key:     "new-checkout",
		Enabled: true,
		TargetingRules: []TargetingRule{
			{
				Attribute: "country",
				Operator:  "equals",
				Value:     "ID",
			},
		},
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id":      "user_123",
		"country": "ID",
	}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonMatchedRule {
		t.Fatalf("expected reason %s, got %s", ReasonMatchedRule, result.Reason)
	}
}

func TestEvaluateEmailEndsWithRuleReturnsTrue(t *testing.T) {
	flag := Flag{
		Key:     "new-checkout",
		Enabled: true,
		TargetingRules: []TargetingRule{
			{
				Attribute: "email",
				Operator:  "ends_with",
				Value:     "@company.com",
			},
		},
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id":    "user_123",
		"email": "alice@company.com",
	}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonMatchedRule {
		t.Fatalf("expected reason %s, got %s", ReasonMatchedRule, result.Reason)
	}
}

func TestEvaluateNotEqualsRuleReturnsTrue(t *testing.T) {
	flag := Flag{
		Key:     "new-checkout",
		Enabled: true,
		TargetingRules: []TargetingRule{
			{
				Attribute: "country",
				Operator:  "not_equals",
				Value:     "SG",
			},
		},
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id":      "user_123",
		"country": "ID",
	}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonMatchedRule {
		t.Fatalf("expected reason %s, got %s", ReasonMatchedRule, result.Reason)
	}
}

func TestEvaluateContainsRuleReturnsTrue(t *testing.T) {
	flag := Flag{
		Key:     "beta-dashboard",
		Enabled: true,
		TargetingRules: []TargetingRule{
			{
				Attribute: "email",
				Operator:  "contains",
				Value:     "alice",
			},
		},
	}

	result := Evaluate(&flag, "beta-dashboard", map[string]any{
		"id":    "user_123",
		"email": "alice@example.com",
	}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonMatchedRule {
		t.Fatalf("expected reason %s, got %s", ReasonMatchedRule, result.Reason)
	}
}

func TestEvaluateMissingUserAttributeDoesNotMatchRule(t *testing.T) {
	flag := Flag{
		Key:     "new-checkout",
		Enabled: true,
		TargetingRules: []TargetingRule{
			{
				Attribute: "country",
				Operator:  "equals",
				Value:     "ID",
			},
		},
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id": "user_123",
	}, false)

	if result.Enabled {
		t.Fatalf("expected enabled false")
	}

	if result.Reason != ReasonDefaultRule {
		t.Fatalf("expected reason %s, got %s", ReasonDefaultRule, result.Reason)
	}
}

func TestEvaluatePercentageRolloutIsDeterministic(t *testing.T) {
	flag := Flag{
		Key:               "new-checkout",
		Enabled:           true,
		RolloutPercentage: 10,
	}

	user := map[string]any{
		"id": "user_123",
	}

	first := Evaluate(&flag, "new-checkout", user, false)
	second := Evaluate(&flag, "new-checkout", user, false)

	if first.Enabled != second.Enabled {
		t.Fatalf("expected deterministic rollout result")
	}

	if first.Reason != second.Reason {
		t.Fatalf("expected deterministic rollout reason")
	}
}
func TestEvaluateHundredPercentRolloutEnablesUser(t *testing.T) {
	flag := Flag{
		Key:               "new-checkout",
		Enabled:           true,
		RolloutPercentage: 100,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id": "user_123",
	}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonPercentageRollout {
		t.Fatalf("expected reason %s, got %s", ReasonPercentageRollout, result.Reason)
	}
}

func TestEvaluatePercentageRolloutWithoutUserIDReturnsDefaultValue(t *testing.T) {
	flag := Flag{
		Key:               "new-checkout",
		Enabled:           true,
		RolloutPercentage: 10,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{}, false)

	if result.Enabled {
		t.Fatalf("expected enabled false")
	}

	if result.Reason != ReasonDefaultRule {
		t.Fatalf("expected reason %s, got %s", ReasonDefaultRule, result.Reason)
	}
}

func TestEvaluateZeroPercentRolloutUsesGlobalEnabled(t *testing.T) {
	flag := Flag{
		Key:               "new-checkout",
		Enabled:           true,
		RolloutPercentage: 0,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id": "user_123",
	}, false)

	if !result.Enabled {
		t.Fatalf("expected enabled true")
	}

	if result.Reason != ReasonDefaultRule {
		t.Fatalf("expected reason %s, got %s", ReasonDefaultRule, result.Reason)
	}
}

func TestEvaluatePercentageRolloutIncludesBucketAndRolloutPercentage(t *testing.T) {
	flag := Flag{
		Key:               "new-checkout",
		Enabled:           true,
		RolloutPercentage: 10,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id": "user_123",
	}, false)

	if result.Bucket == nil {
		t.Fatalf("expected bucket to be present")
	}

	if *result.Bucket < 0 || *result.Bucket > 99 {
		t.Fatalf("expected bucket between 0 and 99, got %d", *result.Bucket)
	}

	if result.RolloutPercentage == nil {
		t.Fatalf("expected rollout percentage to be present")
	}

	if *result.RolloutPercentage != 10 {
		t.Fatalf("expected rollout percentage 10, got %d", *result.RolloutPercentage)
	}
}

func TestEvaluateDisabledFlagDoesNotIncludeRolloutDetails(t *testing.T) {
	flag := Flag{
		Key:               "new-checkout",
		Enabled:           false,
		RolloutPercentage: 10,
	}

	result := Evaluate(&flag, "new-checkout", map[string]any{
		"id": "user_123",
	}, false)

	if result.Bucket != nil {
		t.Fatalf("expected bucket to be nil")
	}

	if result.RolloutPercentage != nil {
		t.Fatalf("expected rollout percentage to be nil")
	}
}
